tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithFunArrays.scala

```diff
@@ -0,0 +1,87 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt.rewriter.ConstSimplifierForSmt
+import at.forsyte.apalache.tla.bmcmt.types._
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, RewriterException, SymbState, SymbStateRewriter}
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.oper.TlaFunOper
+import at.forsyte.apalache.tla.lir.{BoolT1, FunT1, OperEx, TlaEx}
+
+/**
+ * Rewrites set membership tests: x \in S, x \in SUBSET S, and x \in [S -> T].
+ *
+ * @author
+ *   Rodrigo Otoni
+ */
+class SetInRuleWithFunArrays(rewriter: SymbStateRewriter) extends SetInRule(rewriter) {
+  new ConstSimplifierForSmt
+
+  override protected def funSetIn(state: SymbState, funsetCell: ArenaCell, funCell: ArenaCell): SymbState = {
+    // checking whether f \in [S -> T]
+    def flagTypeError(): Nothing = {
+      val msg = s"Not implemented (open an issue): f \\in S for f: %s and S: %s."
+        .format(funCell.cellType, funsetCell.cellType)
+      throw new RewriterException(msg, state.ex)
+    }
+
+    funCell.cellType match {
+      case CellTFrom(FunT1(domT, _)) => domT // OK
+      case _                         => flagTypeError()
+    }
+    funsetCell.cellType match {
+      case FinFunSetT(PowSetT(_), _) | FinFunSetT(FinFunSetT(_, _), _) =>
+        flagTypeError()
+      case _ => () // OK
+    }
+
+    val funDom = state.arena.getDom(funCell)
+    val funsetDom = state.arena.getDom(funsetCell)
+    val funsetCdm = state.arena.getCdm(funsetCell)
+    var nextState = state
+
+    def onPair(arg: ArenaCell): TlaEx = {
+      val funApp = OperEx(TlaFunOper.app, funCell.toNameEx, arg.toNameEx)
```

Comment:
```suggestion
      val funApp = tla.app(funCell.toBuilder, arg.toBuilder)
```

Reply:
There is no `tla.app`. Am I missing something?

Reply:
https://github.com/informalsystems/apalache/blob/1c408463643e32ab5c85dcc606aa6bc7ca88b0dd/tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/FunBuilder.scala#L80

Reply:
@Kukovec I was still using the old builder, so changing `import at.forsyte.apalache.tla.lir.convenience.tla` to `import at.forsyte.apalache.tla.types.tla` allows me to use `tla.app`. The problem is that this began a chain of changes in many files, from using `TlaEx` to using `TBuilderInstruction`. I think it is better to make this change in a separate PR. Thoughts?

Reply:
The thing is, that it should ideally be just as simple as `at.forsyte.apalache.tla.types._`, because the `TBuilderInstruction -> TlaEx` conversion happens automatically. Can you try this, and if it doesn't work in 10min, ignore this comment and go ahead with the old builder.

Reply:
You might need to wrap the `rewriter.lazyEq.cachedEq` in `tla.unchecked` though.

Reply:
Not sure if I'm doing something wrong, but there is still a type mismatch. Another thing is that some operators changed their signature from the old builder to the new one, e.g., `apalacheSelectInSet` to `selectInSet`. Should I go ahead and merge this PR as is?

Reply:
Using `tla.unchecked` fixes individual problems btw, but I don't think we should be using it extensively.

Reply:
Yeah, like I said, if it's not trivial and seamless just go ahead with this and we-ll revisit it once we start the builder transition later